pkg/sql/ttl/ttlbase/ttl_helpers.go

```diff
@@ -12,42 +12,161 @@ package ttlbase
+var endKeyCompareOps = map[catenumpb.IndexColumn_Direction]string{
+	catenumpb.IndexColumn_ASC:  "<",
+	catenumpb.IndexColumn_DESC: ">",
+}
+
+type QueryBounds struct {
+	Start, End tree.Datums
+}
+
+func BuildSelectQuery(
+	relationName string,
+	pkColNames []string,
+	pkColDirs []catenumpb.IndexColumn_Direction,
+	aostDuration time.Duration,
+	ttlExpr catpb.Expression,
+	queryBounds QueryBounds,
+	limit int64,
+	startIncl bool,
+) string {
+	var buf bytes.Buffer
+	// SELECT
+	buf.WriteString("SELECT ")
+	for i := range pkColNames {
 		if i > 0 {
-			b.WriteString(", ")
+			buf.WriteString(", ")
+		}
+		buf.WriteString(pkColNames[i])
+	}
+	// FROM
+	buf.WriteString("\nFROM ")
+	buf.WriteString(relationName)
+	// AS OF SYSTEM TIME
+	buf.WriteString("\nAS OF SYSTEM TIME INTERVAL '")
+	buf.WriteString(strconv.Itoa(int(aostDuration.Milliseconds()) / 1000))
+	buf.WriteString(" seconds'")
+	// WHERE
+	buf.WriteString("\nWHERE ((")
+	buf.WriteString(string(ttlExpr))
+	buf.WriteString(") <= $1)")
+	writeBounds := func(
+		key tree.Datums,
+		placeholderOffset int,
+		compareOps map[catenumpb.IndexColumn_Direction]string,
+		inclusive bool,
+	) {
+		lenKey := len(key)
+		if lenKey > 0 {
+			buf.WriteString("\nAND (")
+			for i := 0; i < lenKey; i++ {
+				isLast := i == lenKey-1
+				buf.WriteString("\n  (")
+				for j := 0; j < i; j++ {
+					buf.WriteString(pkColNames[j])
+					buf.WriteString(" = $")
+					buf.WriteString(strconv.Itoa(j + placeholderOffset))
+					buf.WriteString(" AND ")
+				}
+				buf.WriteString(pkColNames[i])
+				buf.WriteString(" ")
+				buf.WriteString(compareOps[pkColDirs[i]])
+				if isLast && inclusive {
+					buf.WriteString("=")
+				}
+				buf.WriteString(" $")
+				buf.WriteString(strconv.Itoa(i + placeholderOffset))
+				buf.WriteString(")")
+				if !isLast {
+					buf.WriteString(" OR")
+				}
+			}
+			buf.WriteString("\n)")
+		}
+	}
+	const endPlaceholderOffset = 2
+	endPK := queryBounds.End
+	writeBounds(
+		queryBounds.Start,
+		endPlaceholderOffset+len(endPK),
+		startKeyCompareOps,
+		startIncl,
+	)
+	writeBounds(
+		endPK,
+		endPlaceholderOffset,
+		endKeyCompareOps,
+		false, /*inclusive*/
+	)
+
+	// ORDER BY
+	buf.WriteString("\nORDER BY ")
+	for i := range pkColNames {
+		if i > 0 {
+			buf.WriteString(", ")
+		}
+		buf.WriteString(pkColNames[i])
+		buf.WriteString(" ")
+		buf.WriteString(pkColDirs[i].String())
+	}
+	// LIMIT
+	buf.WriteString("\nLIMIT ")
+	buf.WriteString(strconv.Itoa(int(limit)))
+	return buf.String()
+}
+
+func BuildDeleteQuery(
+	relationName string, pkColNames []string, ttlExpr catpb.Expression, numRows int,
+) string {
+	var buf bytes.Buffer
+	// DELETE
+	buf.WriteString("DELETE FROM ")
+	buf.WriteString(relationName)
+	// WHERE
+	buf.WriteString("\nWHERE ((")
+	buf.WriteString(string(ttlExpr))
+	buf.WriteString(") <= $1)")
+	if numRows > 0 {
+		buf.WriteString("\nAND (")
+		for i := range pkColNames {
+			if i > 0 {
+				buf.WriteString(", ")
+			}
+			buf.WriteString(pkColNames[i])
```

Comment:
i believe you need to escape these using `lex.EncodedRestrictSQLIdent`, or for extra mark, let's re-create the string using the `tree.Delete` object instead.

this seems otherwise broken when you have a pk like `"SPECIAL_name" id INT PRIMARY KEY`

Reply:
This now happens here https://github.com/cockroachdb/cockroach/pull/101869/files#diff-a21f73ec9c878a961dc5a592d33403e5d2fabd925c97be208ee6b9df645b504fR114 right after getting the `pkColNames` from the descriptor since everywhere after that might need the escaped version.